### PR TITLE
[To rel/0.12][IOTDB-2528] MLog corruption and recovery bug after killing system

### PR DESCRIPTION
**Pull request by HeimingZ**
Timed force mlog and truncate broken logs.

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/writelog/io/SingleFileLogReader.java [hunk 1]
```diff
@@ -139,4 +143,12 @@ public void open(File logFile) throws FileNotFoundException {
   public boolean isFileCorrupted() {
     return fileCorrupted;
   }
+
+  private void truncateBrokenLogs() {
+    try (FileChannel channel = new FileOutputStream(filepath, true).getChannel()) {
```
Comment (wangchao316) [inline server/src/main/java/org/apache/iotdb/db/writelog/io/SingleFileLogReader.java:148]: channel close,  but FileOutputStream does not close.

File: server/src/main/java/org/apache/iotdb/db/metadata/MManager.java [hunk 2]
```diff
@@ -184,6 +184,17 @@ protected MManager() {
           MTREE_SNAPSHOT_THREAD_CHECK_TIME,
           TimeUnit.SECONDS);
     }
+
+    if (config.getSyncMlogPeriodInMs() != 0) {
+      timedForceMLogThread =
```
Comment (wangchao316) [inline server/src/main/java/org/apache/iotdb/db/metadata/MManager.java:189]: if config.getSyncMlogPeriodInMs() == 0, what will we do ?

File: server/src/main/java/org/apache/iotdb/db/writelog/io/SingleFileLogReader.java [hunk 3]
```diff
@@ -23,13 +23,9 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.io.BufferedInputStream;
-import java.io.DataInputStream;
-import java.io.File;
-import java.io.FileInputStream;
-import java.io.FileNotFoundException;
-import java.io.IOException;
+import java.io.*;
```
Comment (wangchao316) [inline server/src/main/java/org/apache/iotdb/db/writelog/io/SingleFileLogReader.java:26]: does not  import *

**Review by wangchao316 — CHANGES_REQUESTED** [review]

File: docs/UserGuide/Appendix/Config-Manual.md [hunk 4]
```diff
@@ -400,6 +400,24 @@ The permission definitions are in ${IOTDB\_CONF}/conf/jmx.access.
...
+|Default| 1048576 |
+|Effective|After restart system|
+
+* sync\_mlog\_period\_in\_ms
+
+|Name| sync\_mlog\_period\_in\_ms |
+|:---:|:---|
+|Description| The cycle when metadata log is periodically forced to be written to disk(in milliseconds)|
+|Type| Int64 |
+|Default| 0 |
```
[the remark this answers is not in the document]
Comment (HeimingZ) [reply docs/UserGuide/Appendix/Config-Manual.md:418]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/metadata/MManager.java (hunk 2 above)
Comment (HeimingZ) [reply server/src/main/java/org/apache/iotdb/db/metadata/MManager.java:189]: Sync mlog after each refreshment, you can see it in the constructors of MLogWriter.

File: server/src/main/java/org/apache/iotdb/db/writelog/io/SingleFileLogReader.java (hunk 3 above)
Comment (HeimingZ) [reply server/src/main/java/org/apache/iotdb/db/writelog/io/SingleFileLogReader.java:26]: They are modified by spotless, is there any method to skip these modifications?

File: server/src/main/java/org/apache/iotdb/db/writelog/io/SingleFileLogReader.java (hunk 1 above)
Comment (HeimingZ) [reply server/src/main/java/org/apache/iotdb/db/writelog/io/SingleFileLogReader.java:148]: Change code like this?
`
try (FileOutputStream outputStream = new FileOutputStream(filepath, true);
        FileChannel channel = outputStream.getChannel()) {
`

Comment (wangchao316) [reply server/src/main/java/org/apache/iotdb/db/writelog/io/SingleFileLogReader.java:148]: yes

File: server/src/test/java/org/apache/iotdb/db/writelog/io/LogWriterReaderTest.java [hunk 5]
```diff
@@ -94,4 +97,102 @@ public void testWriteAndRead() throws IOException {
...
+      while (reader.hasNext()) {
+        res.add(reader.next());
+      }
+      for (int i = 0; i < plans.size(); i++) {
+        assertEquals(plans.get(i), res.get(i));
+      }
+      reader.close();
+      assertEquals(expectedLength, new File(filePath).length());
+    } finally {
+      new File(filePath).delete();
```
Comment (wangchao316) [inline server/src/test/java/org/apache/iotdb/db/writelog/io/LogWriterReaderTest.java:141]: please move "channel.close(), reader.close()"  to "finally". 
If the test case is abnormal and the connection is not closed, the execution of other test cases will be affected.

**Review by wangchao316 — CHANGES_REQUESTED** [review]

File: server/src/test/java/org/apache/iotdb/db/writelog/io/LogWriterReaderTest.java (hunk 5 above)
Comment (HeimingZ) [reply server/src/test/java/org/apache/iotdb/db/writelog/io/LogWriterReaderTest.java:148]: Fixed.

File: server/src/main/java/org/apache/iotdb/db/writelog/io/SingleFileLogReader.java (hunk 3 above)
Comment (HeimingZ) [reply server/src/main/java/org/apache/iotdb/db/writelog/io/SingleFileLogReader.java]: Fixed.

**Review by wangchao316 — APPROVED** [review]
LGTM

**Review by MarcosZyk — APPROVED** [review]
LGTM~

**Pull request merged**